### PR TITLE
initial commit of cloudsat decoder

### LINKED ISSUE #1498: add Cloudsat ioda converter
## Description

create a Cloudsat Cloud Profiling Radar (CPR)  decoder to format the data to IODA format

## Requirements

python IODA interface

## Acceptance Criteria (Definition of Done)

converter capable of processing data into IODA format

### PR DESCRIPTION
**Pull request by BenjaminRuston**
## Description

Initial commit of a cloudSat decoder for CPR L2 data

## Issue(s) addressed

Resolves #1498 

## Impact

Expected impact on downstream repositories:

## Checklist

- [ ] I have performed a self-review of my own code
- [ ] I have run the unit tests before creating the PR

### CODE REVIEW

**Comment by BenjaminRuston** [comment]
@imoradi this is just initial commit, I've removed the dependency on `jdutil` and just computed the epoch time directly

there is still some cleanup yet to do but can you verify this still works for you.

**Comment by imoradi** [comment]
Other than the way you calculate epoch which I cannot confirm because I don't have JEDI Python tools, everything else works. I can run and generate the output simply by setting epoch to a specific date.

**Comment by BenjaminRuston** [comment]
@imoradi I can show you how to run this independently of the pyiodaconv library (quick hack) but also will show you how to build these and run converter on a NASA machine

File: src/hdf5/read_cloudsat.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+import copy
+
+import numpy as np
+import datetime
+import pdb
+import xarray as xr
+from pyiodaconv.def_jedi_utils import iso8601_string
+os.environ["HDF5_USE_FILE_LOCKING"] = "FALSE"
+
+epoch = datetime.datetime.fromisoformat(iso8601_string[14:-1])
```
Comment (BenjaminRuston) [inline src/hdf5/read_cloudsat.py]: @imoradi you comment out:
```
from pyiodaconv.def_jedi_utils import iso8601_string
```

and then get the `iso8601_string` from here:

https://github.com/JCSDA-internal/ioda-converters/blob/0bc00802515fe5834fb70311bfcac3f4cf12d349/src/pyiodaconv/def_jedi_utils.py#L20

Comment (imoradi) [reply src/hdf5/read_cloudsat.py]: Yes it works fine

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+import datetime
+import time
+import read_cloudsat
+
+# fname_in = ['2009335004815_19116_CS_2B-GEOPROF_GRANULE_P1_R05_E02_F00.hdf']
+fname_in = ['2009212223327_17338_CS_2B-GEOPROF_GRANULE_P1_R05_E02_F00.hdf']
+fname_out = "cpr_cloudsat_obs_2009335.nc4"
+
+cpr_obs = read_cloudsat.read_cloudsat(fname_in)
+cpr_obs["sequenceNumber"] = xr.DataArray(np.arange(cpr_obs.obs_id.size), cpr_obs.obs_id.coords)
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py]: @imoradi would be good to discuss this with you

why isn't this particular specification of `sequenceNumber` in the `read_cloudsat`?   But think i've got a handle on this and would like to pick it up if you're OK with that

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+    # Groups ['GsiHofX', 'MetaData', 'ObsError', 'ObsValue', 'PreQC', 'RecMetaData']
+
+    g = 'ObsValue'
+    ds_g = ds.createGroup(g)
+    zvar =  ds_g.createVariable('ReflectivityAttenuated', np.float32, refl_dims, fill_value=fill_value_dbz, zlib=True)
+    ds_g.variables['ReflectivityAttenuated'][:] = cpr_obs.obs.values.astype(np.float32)
+    zvar.setncattr_string("units", str("dBz"))
+    
+    
+    g = 'GsiHofX'
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py]: These are non-standard groups and do not exist in the observational data.  

More specifically `Gsi` is obviously not generic across partners so this decoder should function just to decode raw data files

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+        ds_g.createVariable('sensorViewAngle', np.float32, ('Location',), fill_value=fill_value)
+        ds_g.createVariable('sensorZenithAngle', np.float32, ('Location',), fill_value=fill_value)
+        ds_g.createVariable('sensorAzimuthAngle', np.float32, ('Location',), fill_value=fill_value)
+        ds_g.createVariable('solarAzimuthAngle', np.float32, ('Location',), fill_value=fill_value)
+        ds_g.createVariable('solarZenithAngle', np.float32, ('Location',), fill_value=fill_value)
+        # ds_g.createVariable('time', np.float32, ('Location',), fill_value=fill_value)
+        ds_g.createVariable('sequenceNumber', np.int32, ('Location',), fill_value=-2147483647)
+
+        ds_g.variables['sensorCentralFrequency'][:] = np.array([94.05]).astype(np.float32)    # cpr_sim.Frequency.values.astype(np.float32)
+        ds_g.variables['sensorCentralWavenumber'][:] = np.array([3.1371]).astype(np.float32)  # cpr_sim.Wavenumber.values.astype(np.float32)
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py]: @imoradi these lines make me nervous is this correct

I will look into it some more in a bit

**Comment by BenjaminRuston** [comment]
> Other than the way you calculate epoch which I cannot confirm because I don't have JEDI Python tools, everything else works. I can run and generate the output simply by setting epoch to a specific date.

@imoradi  I can show you how to make a the common util available (it just lives in a build folder rather than the src folder)

**Comment by imoradi** [comment]
The values are correct but we have to find a better way than hardcoding them

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    VarAttrs = DefaultOrderedDict(lambda: DefaultOrderedDict(dict))
+    set_obspace_attributes(VarAttrs)
+    set_metadata_attributes(VarAttrs)
+
+    VarAttrs[(k, 'ObsValue')]['_FillValue'] = float_missing_value
+    VarAttrs[(k, 'ObsError')]['_FillValue'] = float_missing_value
+    VarAttrs[(k, 'PreQC')]['_FillValue'] = int_missing_value
+    VarAttrs[(k, 'ObsValue')]['units'] = 'dBz'
+    VarAttrs[(k, 'ObsError')]['units'] = 'dBz'
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py:93]: @imoradi are these the correct units?

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+
+    obs_data[('latitude', metaDataName)] = cpr_obs.lat.values.astype(np.float32)
+    obs_data[('longitude', metaDataName)] = cpr_obs.lon.values.astype(np.float32)
+    obs_data[('satelliteIdentifier', metaDataName)] = np.full((nobs), WMO_sat_ID, dtype='int32')
+    obs_data[('sensorCentralFrequency', metaDataName)] = np.array([94.05]).astype(np.float32)    # cpr_sim.Frequency.values.astype(np.float32)
+    obs_data[('sensorCentralWavenumber', metaDataName)] = np.array([3.1371]).astype(np.float32)  # cpr_sim.Wavenumber.values.astype(np.float32)
+    obs_data[('sensorPolarizationDirection', metaDataName)] = np.array([9]).astype(np.int32)
+    obs_data[('sensorScanPosition', metaDataName)] = cpr_obs.fov1.values.astype(np.int32)
+    obs_data[('sensorChannelNumber', metaDataName)] = cpr_obs.channel.values.astype(np.int32)
+    obs_data[('sensorViewAngle', metaDataName)] = cpr_obs.zenith_angle.values.astype(np.float32)
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py:139]:
```suggestion
    # use compute_scan_angle need the satellite altitude in meters
    # 
    # obs_data[('sensorViewAngle', metaDataName)] = compute_scan_angle(
    # cpr.obs.zenith_angle.values,
    #   satellite_altitude)
```

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    obs_data[('sensorScanPosition', metaDataName)] = cpr_obs.fov1.values.astype(np.int32)
+    obs_data[('sensorChannelNumber', metaDataName)] = cpr_obs.channel.values.astype(np.int32)
+    # add satellite altitude (height in meters) and use compute_scan_angle function
+    obs_data[('sensorViewAngle', metaDataName)] = cpr_obs.zenith_angle.values.astype(np.float32)
+    obs_data[('sensorZenithAngle', metaDataName)] = cpr_obs.zenith_angle.values.astype(np.float32)
+    obs_data[('sensorAzimuthAngle', metaDataName)] = cpr_obs.azimuth_angle.values.astype(np.float32)
+    obs_data[('solarZenithAngle', metaDataName)] = np.zeros(nobs).astype(np.float32)
+    obs_data[('solarAzimuthAngle', metaDataName)] = np.zeros(nobs).astype(np.float32)
+    obs_data[('sequenceNumber', metaDataName)] = cpr_obs.sequenceNumber.values.astype(np.int32)
+    # ?? ds_g.variables['sequenceNumber'][:] = cpr_obs["sequenceNumber"].values.astype(np.int32)
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py]:
```suggestion
```

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+float_missing_value = iconv.get_default_fill_val(np.float32)
+int_missing_value = iconv.get_default_fill_val(np.int32)
+long_missing_value = iconv.get_default_fill_val(np.int64)
+
+metaDataName = iconv.MetaDataName()
+obsValName = iconv.OvalName()
+
+# globals
+CLOUDSAT_WMO_sat_ID = 788
+
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py:34]: add satellite height here too for now... these should be moved into general utility shortly

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+    WMO_sat_ID = get_WMO_satellite_ID('CloudSat')
+
+    # allocate space for output depending on which variables are to be saved
+    obs_data = init_obs_loc()
+
+    obs_data[('latitude', metaDataName)] = cpr_obs.lat.values.astype(np.float32)
+    obs_data[('longitude', metaDataName)] = cpr_obs.lon.values.astype(np.float32)
+    obs_data[('satelliteIdentifier', metaDataName)] = np.full((nobs), WMO_sat_ID, dtype='int32')
+    obs_data[('sensorCentralFrequency', metaDataName)] = np.array([94.05]).astype(np.float32)    # cpr_sim.Frequency.values.astype(np.float32)
+    obs_data[('sensorCentralWavenumber', metaDataName)] = np.array([3.1371]).astype(np.float32)  # cpr_sim.Wavenumber.values.astype(np.float32)
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py:128]: @imoradi these too assume the RADAR has a single channel and the frequency is fixed ?

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+    obs_data[(k, "PreQC")] = np.full((nobs, nchans), 0, dtype='int32')
+
+    return obs_data
+
+
+def init_obs_loc():
+    obs = {
+        ('ReflectivityAttenuated', "ObsValue"): [],
+        ('ReflectivityAttenuated', "ObsError"): [],
+        ('ReflectivityAttenuated', "PreQC"): [],
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py]:
```suggestion
    k = radarReflectivityAtt
    obs = {
        (k, "ObsValue"): [],
        (k, "ObsError"): [],
        (k, "PreQC"): [],
```

File: src/hdf5/read_cloudsat.py [hunk 11]
```diff
@@ -0,0 +1,209 @@
...
+    csdata['lat'] = np.squeeze(latitude)
+    csdata['lon'] = np.squeeze(longitude)
+    csdata['height'] = np.squeeze(height)
+    csdata['dem_elevation'] = np.squeeze(dem)
+    csdata['surface_height'] = np.squeeze(surface_height)
+    csdata['SurfaceHeightBin'] = np.squeeze(SurfaceHeightBin)
+    csdata['epoch_time'] = np.squeeze(epoch_time)
+    csdata['obs'] = np.squeeze(dataf)
+    csdata['zenith_angle'] = csdata['lat'].copy() * 0
+    csdata['azimuth_angle'] = csdata['lat'].copy() * 0
```
Comment (BenjaminRuston) [inline src/hdf5/read_cloudsat.py:151]: @imoradi am I correct that this looks like we are not getting a `zenith` or `azimuth` angle for the observations?

File: src/hdf5/cloudsat_netcdf_to_ioda.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+    obs_data[('sensorCentralWavenumber', metaDataName)] = np.array([3.1371]).astype(np.float32)  # cpr_sim.Wavenumber.values.astype(np.float32)
+    obs_data[('sensorPolarizationDirection', metaDataName)] = np.array([9]).astype(np.int32)
+    obs_data[('sensorScanPosition', metaDataName)] = cpr_obs.fov1.values.astype(np.int32)
+    obs_data[('sensorChannelNumber', metaDataName)] = cpr_obs.channel.values.astype(np.int32)
+    # add satellite altitude (height in meters) and use compute_scan_angle function
+    obs_data[('sensorViewAngle', metaDataName)] = cpr_obs.zenith_angle.values.astype(np.float32)
+    obs_data[('sensorZenithAngle', metaDataName)] = cpr_obs.zenith_angle.values.astype(np.float32)
+    obs_data[('sensorAzimuthAngle', metaDataName)] = cpr_obs.azimuth_angle.values.astype(np.float32)
+    obs_data[('solarZenithAngle', metaDataName)] = np.zeros(nobs).astype(np.float32)
+    obs_data[('solarAzimuthAngle', metaDataName)] = np.zeros(nobs).astype(np.float32)
```
Comment (BenjaminRuston) [inline src/hdf5/cloudsat_netcdf_to_ioda.py:137]: @imoradi we can populate zenith and azimuth to zero here too if needed

**Comment by BenjaminRuston** [comment]
@imoradi your change to add `Layer` and `height` should be in here now

**Review by imoradi — APPROVED** [review]
Looks good!

**Comment by BenjaminRuston** [comment]
while I do a small flurry of commits, turning off CI and resetting to draft

a ctest is being added to test this converter

**Comment by BenjaminRuston** [comment]
@fcvdb can you please have a look and we can move this forward there is another PR in draft that will add GPM DPR using this as template

**Review by fcvdb — APPROVED** [review]
Thanks Ben!

**Pull request merged**